### PR TITLE
[sdk/python] improved vectors generation

### PR DESCRIPTION
**Pull request by gimre-xymcity**
## What's the issue?
There's a lot of repetitions and lot of senseless cases (mostly) in symbol tx test vectors

## How have you changed the behavior?
 * converted "generate" functions to "recipes
 * removed duplicates
 * removed senseless cases
 * introduced new cases, mostly oriented around flags covering

## How was this change tested?
 * previous change (#99 verified that generator actually emits proper payloads,
 * this change introduces new vectors, existing (py) vector tester passes on all of them (haven't checked js, should pass as well)

### CODE REVIEW

File: sdk/python/testvectors/README.md [hunk 1]
```diff
@@ -1,5 +1,3 @@
 # testvectors
 
 This directory contains testvectors generator written from scratch based on pre-existing testvectors.
```
Comment (Jaguar0625) [inline sdk/python/testvectors/README.md:3]: not sure if we need this readme?

File: sdk/python/testvectors/__main__.py [hunk 2]
```diff
@@ -1,28 +1,31 @@
...
+		'nem': [
+			'account_key_link', 'cosignature', 'mosaic_definition', 'mosaic_supply_change', 'multisig',
+			'multisig_account_modification_v1', 'multisig_account_modification', 'namespace_registration', 'transfer_v1', 'transfer'],
+		'symbol': [
+			'account_address_restriction', 'account_key_link', 'account_metadata', 'account_mosaic_restriction',
+			'account_operation_restriction', 'address_alias', 'aggregate_bonded', 'aggregate_complete', 'hash_lock',
+			'mosaic_address_restriction', 'mosaic_alias', 'mosaic_definition', 'mosaic_global_restriction', 'mosaic_metadata',
+			'mosaic_supply_change', 'multisig_account_modification', 'namespace_metadata', 'namespace_registration', 'node_key_link',
+			'secret_lock', 'secret_proof', 'transfer', 'voting_key_link', 'vrf_key_link']
 	}
```
Comment (Jaguar0625) [inline sdk/python/testvectors/__main__.py:29]: any reason not to autodetect ?

File: sdk/python/testvectors/__main__.py [hunk 3]
```diff
@@ -33,44 +36,225 @@ def get_modules(network_name):
...
+
+class SymbolHelper:
+	AGREGATE_SCHEMA_NAME = 'AggregateBondedTransaction'
+	Signature = sc.Signature
+
+	def __init__(self, network_name):
+		self.facade = SymbolFacade(network_name)
+
+	def set_common_fields(self, descriptor, test_name):
+		descriptor['signer_public_key'] = sha3.sha3_256(test_name.encode('utf8')).hexdigest()
```
Comment (Jaguar0625) [inline sdk/python/testvectors/__main__.py:63]: i guess this doesn't need to be wrapped in PublicKey?

File: sdk/python/testvectors/__main__.py [hunk 4]
```diff
@@ -33,44 +36,225 @@ def get_modules(network_name):
...
+
+	def create_aggregate(self, test_name, original_descriptor):
+		factory = self.facade.transaction_factory
+		embedded_transactions = list(map(clone_descriptor, original_descriptor['embedded']))
+
+		printable_descriptor = clone_descriptor(original_descriptor['aggregate'])
+		printable_descriptor['transactions'] = embedded_transactions
+		self.set_common_fields(printable_descriptor, test_name)
+		printable_descriptor['cosignatures'] = [
+			self.create_cosignature_descriptor(test_name, i) for i in range(original_descriptor.get('num_cosignatures', 0))]
```
Comment (Jaguar0625) [inline sdk/python/testvectors/__main__.py:102]: i think more readable for closing ] to be on own line

File: sdk/python/testvectors/__main__.py [hunk 5]
```diff
@@ -33,44 +36,225 @@ def get_modules(network_name):
...
+class NemHelper:
+	AGREGATE_SCHEMA_NAME = 'MultisigTransaction'
+	Signature = nc.Signature
+
+	def __init__(self, network_name):
+		self.facade = NemFacade(network_name)
+
+	def set_common_fields(self, descriptor, test_name):
+		descriptor['signature'] = self.Signature(sha3.sha3_512(test_name.encode('utf8')).hexdigest())
+		descriptor['signer_public_key'] = sha3.sha3_256(test_name.encode('utf8')).hexdigest()
```
Comment (Jaguar0625) [inline sdk/python/testvectors/__main__.py:123]: swap order to match SymbolHelper order?

File: sdk/python/testvectors/__main__.py [hunk 6]
```diff
@@ -33,44 +36,225 @@ def get_modules(network_name):
...
 class VectorGenerator:
 	def __init__(self, network_name):
-		test_cases = prepare_transaction_test_cases(network_name)
 		self.network_name = network_name
 		self.modules = get_modules(network_name)
-		self.cases = {case['test_name']: case for case in test_cases}
-		facade_class = {'symbol': SymbolFacade, 'nem': NemFacade}[network_name]
-		self.facade = facade_class('testnet')
+		helper_class = {'symbol': SymbolHelper, 'nem': NemHelper}[network_name]
+		self.helper = helper_class('testnet')
```
Comment (Jaguar0625) [inline sdk/python/testvectors/__main__.py:175]: self.network_name ?

File: sdk/python/testvectors/__main__.py [hunk 7]
```diff
@@ -33,44 +36,225 @@ def get_modules(network_name):
...
+
+	def create_transactions(self, module_descriptor, recipes):
+		test_cases = []
+		schema_name = recipes['schema_name']
+		test_prefix = f'{schema_name}_{module_descriptor[0]}'
+		for index, descriptor in enumerate(recipes['descriptors']):
+			test_name = f'{test_prefix}_single_{index+1}'
+			test_cases.append(self.create_entry(schema_name, test_name, self.helper.create, descriptor))
+
+		# only thing _not_ supporting wrapping in aggregates is NEM's Cosignature (transaction)``
```
Comment (Jaguar0625) [inline sdk/python/testvectors/__main__.py:210]: extra `` ?

File: sdk/python/testvectors/__main__.py [hunk 8]
```diff
@@ -33,44 +36,225 @@ def get_modules(network_name):
...
+		return entries
+
 
 def main():
 	for network_name in ['symbol', 'nem']:
 		generator = VectorGenerator(network_name)
-		generator.generate()
+		entries = generator.generate()
+
+		filepath = Path(os.environ.get('SCHEMAS_PATH', '.')) / network_name / 'transactions' / 'transactions_new.yaml'
```
Comment (Jaguar0625) [inline sdk/python/testvectors/__main__.py:253]: any reason not to use cmdline arg ?

File: sdk/python/testvectors/__main__.py [hunk 9]
```diff
@@ -33,44 +36,225 @@ def get_modules(network_name):
...
+		with open(filepath, 'wt', encoding='utf8') as output_file:
```
Comment (Jaguar0625) [inline sdk/python/testvectors/__main__.py:254]: outfile ?

File: sdk/python/testvectors/__main__.py (hunk 2 above)
Comment (gimre-xymcity) [reply sdk/python/testvectors/__main__.py:29]: I thought you wanted explicit list?

File: sdk/python/testvectors/__main__.py (hunk 3 above)
Comment (gimre-xymcity) [reply sdk/python/testvectors/__main__.py:63]: nope (although could)

File: sdk/python/testvectors/__main__.py (hunk 2 above)
Comment (Jaguar0625) [reply sdk/python/testvectors/__main__.py:29]: i don't remember that 😅 i think autodetecting should be fine

File: sdk/python/testvectors/README.md (hunk 1 above)
Comment (gimre-xymcity) [reply sdk/python/testvectors/README.md:3]: we probably should have readme for any tool/script (or at least describe usage _somewhere_)

File: sdk/python/testvectors/__main__.py (hunk 8 above)
Comment (gimre-xymcity) [reply sdk/python/testvectors/__main__.py:253]: it started similar to our test scripts

**Comment by Jaguar0625** [comment]
1. lint is failing: https://jenkins.symboldev.com/blue/organizations/jenkins/Symbol%2Fgenerated%2Fsymbol%2Fpython/detail/PR-118/5/pipeline
2. do we need _new file prefix still?

File: sdk/python/tests/facade/test_SymbolFacade.py [hunk 10]
```diff
@@ -204,6 +233,43 @@ def test_can_verify_aggregate_transaction(self):
...
+	def test_can_hash_aggregate_embedded_transactions(self):
+		# Arrange:
+		facade = SymbolFacade('testnet', AccountDescriptorRepository(YAML_INPUT))
+		transaction = self._create_real_aggregate(facade)
+
+		# Act:
+		hash_value = facade.hash_aggregate_embedded_transactions(transaction.transactions)
+
+		# Assert:
+		self.assertEqual(transaction.transactions_hash.bytes, hash_value.bytes)
```
[the remark this answers is not in the document]
Comment (gimre-xymcity) [reply sdk/python/tests/facade/test_SymbolFacade.py:247]: yup, and wasn't sure if I should just wrap one of them so they'd be equal

File: tests/vectors/symbol/transactions/transactions.yaml [hunk 11]
```diff
@@ -1,397 +1,1446 @@
...
+    registration_type: root
+    signature: C8BE2E3FBBCB8B448B30416D079EAD7422BC2C6B27106B8A78EA0BD0FA4EB87750EAA3DC387EAAF9F0D0476194D19F974C320CCB306299BE0A04877C95B361D3
+    signer_public_key: fb36dbbd3944a969969707cc0fe8ddadb58f3fecc7256492d867a36bbca52ea8
+    type: namespace_registration_transaction
+  payload: 9E00000000000000C8BE2E3FBBCB8B448B30416D079EAD7422BC2C6B27106B8A78EA0BD0FA4EB87750EAA3DC387EAAF9F0D0476194D19F974C320CCB306299BE0A04877C95B361D3FB36DBBD3944A969969707CC0FE8DDADB58F3FECC7256492D867A36BBCA52EA80000000001984E41E0FEEEEFFEEEEFFEE0711EE7711EE77110270000000000007EE9B3B8AFDF53C0000C6E65776E616D657370616365
+  schema_name: NamespaceRegistrationTransaction
+  test_name: NamespaceRegistrationTransaction_namespace_registration_single_1
+- descriptor:
+    deadline: 8207562320463688160
+    duration: 10000
```
[the remark this answers is not in the document]
Comment (gimre-xymcity) [reply tests/vectors/symbol/transactions/transactions.yaml:941]: hmm, right, wondering if that is something that transactionfactory could catch...

File: sdk/python/tests/facade/test_SymbolFacade.py [hunk 12]
```diff
@@ -235,24 +235,24 @@ def test_can_verify_aggregate_transaction(self):
 
 	# region aggregate_hash_embedded_transactions
```
Comment (Jaguar0625) [inline sdk/python/tests/facade/test_SymbolFacade.py:236]: update

**Review by Jaguar0625 — APPROVED** [review]

**Comment by gimre-xymcity** [comment]
merged into feature branch

**Pull request merged**